ARTstor.js

```diff
@@ -1,131 +1,899 @@
+                        var key = masterObj.id + ":" + masterObj.type;
+                        candidateItems[key] = masterObj.title;
+                    }
+                }
+                if (zMode && Object.keys(candidateItems).length > 0) {
+                    var objItems = [];
+                    for (var objItem in candidateItems) {
+                        objItems.push(objItem);
+                        processObjects(doc, url, objItems);
+                        break;
+                    }
+                } else {
+                    Zotero.selectItems(candidateItems, function(selectItems) {
+                        var objItems = [];
+                        for (var objItem in selectItems) {
+                            objItems.push(objItem);
+                        }
+                        processObjects(doc, url, objItems);
+                    });
+                }
+                break;
+            case 1: // item from metadata window or image viewer
+                // build look up table for eazy look up.
+                var masterLookup = new Object();
+                for (var m = 0; m < masterObjList.length; m++) {
+                    masterLookup[masterObjList[m].id] = masterObjList[m];
+                }
+                var objItems = [];
+                for (var i = 0; i < selectedObjs.length; i++) {
+                    var id = selectedObjs[i];
+                    if (masterLookup[id] == "") {
+                        var item = id + ":10"; // default type as image.
+                        objItems.push(item);
+                    } else {
+                        var item = masterLookup[id].id + ":" + masterLookup[id].type;
+                        objItems.push(item);
+                    }
+                }
+                processObjects(doc, url, objItems);
+                break;
+        }
+    }); // Zotero..doGet
+};
+
+
+function doThumbnails(doc, url) {
+    var selectedObjs = getSelectedItems(doc, url);
+    processSelectedObject(doc, url, selectedObjs, 2);
+}
+
+/**
+    getMasterThumbnailList gets the thumbnail object id by calling the thumbnail 
+    service to get a list of object id and its title.
+    It returns with a list of object with object id, and object title.
+**/
+function getMasterThumbnailList(doc, url, objDescItems) {
+    // var objDescItems = [];
+    var serviceURL = getThumbnailServiceURL(doc, url);
+    Zotero.Utilities.HTTP.doGet(serviceURL, function(text) {
+        var json = JSON.parse(text);
+        var objDescItem;
+        for (var i = 0; i < json.thumbnails.length; i++) {
+            var thumbnail = json.thumbnails[i];
+            objDescItem = new Object();
+            objDescItem.id = thumbnail.objectId;
+            objDescItem.type = thumbnail.objectTypeId;
+            objDescItem.title = thumbnail.thumbnail1;
+            objDescItems.push(objDescItem);
+        }
+    });
+    return objDescItems;
```

Comment:
Well, not undefined, since you pass it in. But if you're expecting `objDescItems` to have changed after the function returns, then that's not the case. In general, do not do this. Use callbacks.